File: sem_1/EAS520/Assignments/Assignment_1/r_code/p5_vv_prod.R
library(pracma)

n <- 10000
a <- runif(n)
b <- runif(n)
c <- 0


# Dot product with vectorization
t1 <- Sys.time()
cc <- a%*%b
t2 <- Sys.time()

time_vec = t2 - t1

# Dot product with for-loop
t3 <- Sys.time()
for (i in 1:n) {
  c <- c + a[i]*b[i]
}
t4 <- Sys.time()
c
time_loop = t4 - t3

norm <- (cc-c)

speed_up <- as.double(time_loop, units='secs')/as.double(time_vec, units='secs')

n
norm
speed_up

File: EAS520/Assignments/Assignment_1/python_code/p5_vv_prod.py
# Dot Product of 2 vectors

from time import time

import numpy as np

n = 10000  # Length of the vectors

a = np.random.rand(n, 1)  # Create a random column vector a
b = np.random.rand(n, 1)  # Create a random column vector b

c = 0  # pre-allocate memory to store the result in c

# Dot product with for-loop
t1 = time()
for i in range(n):
    c += a[i] * b[i]
t2 = time()

time_loop = t2 - t1


# Dot product with numpy
t3 = time()

cc = np.vdot(a, b)

t4 = time()

time_numpy = t4 - t3


# Compare the results
norm = np.linalg.norm(cc - c)

# Measure the speed-up
speedup = time_loop / time_numpy

print(cc-c)
print("n: ", n)
print("norm: ", norm)
print("Speedup: ", speedup)
